store/tikv/txn.go

```diff
@@ -319,34 +320,28 @@ func (txn *tikvTxn) close() {
 }
 
 func (txn *tikvTxn) Rollback() error {
+	if !txn.valid {
+		return kv.ErrInvalidTxn
+	}
 	// Clean up pessimistic lock.
 	if txn.IsPessimistic() && txn.committer != nil {
-		err := txn.rollbackPessimisticLock()
+		err := txn.rollbackPessimisticLocks()
 		if err != nil {
 			logutil.Logger(context.Background()).Error(err.Error())
 		}
 	}
-
-	if !txn.valid {
-		return kv.ErrInvalidTxn
-	}
 	txn.close()
 	logutil.Logger(context.Background()).Debug("[kv] rollback txn", zap.Uint64("txnStartTS", txn.StartTS()))
 	tikvTxnCmdCountWithRollback.Inc()
 
 	return nil
 }
 
-func (txn *tikvTxn) rollbackPessimisticLock() error {
-	c := txn.committer
-	if err := c.initKeysAndMutations(); err != nil {
-		return errors.Trace(err)
-	}
-	if len(c.keys) == 0 {
+func (txn *tikvTxn) rollbackPessimisticLocks() error {
+	if len(txn.lockKeys) == 0 {
 		return nil
 	}
-
-	return c.cleanupKeys(NewBackoffer(context.Background(), cleanupMaxBackoff), c.keys)
+	return txn.committer.pessimisticRollbackKeys(NewBackoffer(context.Background(), cleanupMaxBackoff), txn.lockKeys)
```

Comment:
Maybe cleanupKeys is more suitable here?
cleanupKeys checks `startTS` (the whole txn) while pessimisticRollbackKeys checks `startTS + forUpdateTS`(only the statement)

Reply:
We use `pessimisticRollbackKeys`, so there will not be rollback records written.

Reply:
The check has changed to `lock.forUpdateTS <= forUpdateTS`.

Reply:
It turns out use PessimisticRollback will not clean up the deadlock detector, so we have to use `cleanupKeys`